#20587 Clean artery configuration

2m opened the pull request:
Fixes #20587

2m committed:
#20587 Move artery settings from remoting settings to dedicated class.

2m committed:
#20587 Move hardcoded settings to configuration file.

patriknw commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -147,41 +180,41 @@ akka {
         compression {
           # possibility to disable compression by setting this to off
           enabled = on
```
I think we should remove this, and only have it as programatic toggle for testing (if we need to turn it off at all). We could use a long `advertisement-interval` for such tests.

ktoso replied:
I agree. We want to have it enabled nowawayd

patriknw commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -147,41 +180,41 @@ akka {
-          
+
           # unlocks additional very verbose debug logging of compression events (on DEBUG log level)
           debug = off
```
do we need this? can it be a compile time constant, if it's only for our own debugging purposes?

ktoso replied:
It can be a compile time constant

patriknw commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -147,41 +180,41 @@ akka {
-          
+
           actor-refs {
             enabled = off # TODO possibly remove on/off option once we have battle proven it?
```
remove

patriknw replied:
this is probably not even used

patriknw commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -147,41 +180,41 @@ akka {
+            advertisement-interval = 1 minute # TODO find good number as default, for benchmarks trigger immediately
           }
           manifests {
             enabled = off # TODO possibly remove on/off option once we have battle proven it?
```
remove

patriknw commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -689,8 +722,19 @@ akka {
+#   akka.remote.artery.advanced = ${artery-profile-low-latency}
+artery-profile-low-latency {
+  idle-cpu-level = 9
+}
```
if we can't find any other properties that should be derived from  "profile" I don't think we should do it

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/ArterySettings.scala:
```diff
@@ -0,0 +1,90 @@
+    val DeleteAeronDirectory = getBoolean("delete-aeron-dir")
+    val IdleCpuLevel: Int = getInt("idle-cpu-level").requiring(level ⇒
+      1 <= level && level <= 10, "idle-cpu-level must be between 1 and 10")
+    val SystemMessageResendInterval = config.getMillisDuration("system-message-resend-interval")
```
add more validation checks with `requiring` where it makes sense, e.g. here > 0

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -306,38 +305,25 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
   // TODO use WildcardIndex.isEmpty when merged from master
   val largeMessageChannelEnabled =
-    !largeMessageDestinations.wildcardTree.isEmpty || !largeMessageDestinations.doubleWildcardTree.isEmpty
+    !remoteSettings.Artery.LargeMessageDestinations.wildcardTree.isEmpty || !remoteSettings.Artery.LargeMessageDestinations.doubleWildcardTree.isEmpty
```
perhaps add a `val settings = remoteSettings.Artery` to reduce code length

patriknw commented:
looking good

ktoso commented on akka-remote/src/main/scala/akka/remote/artery/ArterySettings.scala:
```diff
@@ -0,0 +1,90 @@
+      tree.insert(segments, NotUsed)
+    }
+
+  object Advanced {
```
doesn't this style make it impossible to programatically change values?

See https://github.com/ktoso/kaze-class style and how HTTP deals with settings perhaps?

patriknw replied:
all those are akka internal and we can change them however we want

ktoso commented on akka-remote/src/main/scala/akka/remote/artery/ArterySettings.scala:
```diff
@@ -0,0 +1,90 @@
+    val TestMode: Boolean = getBoolean("test-mode")
+    val MaterializerSettings = ActorMaterializerSettings(config.getConfig("materializer"))
+    val EmbeddedMediaDriver = getBoolean("embedded-media-driver")
+    val AeronDirectoryName = getString("aeron-dir") requiring (dir ⇒
```
if stored in an object, won't that make it impossible to just for fun run 2 akka nodes via remoting in the same jvm, like we do im sample apps sometimes?

ktoso replied:
Ignore that, I see it's a nested object in a class. All good then.

ktoso commented:
Looks good. I'm not sure about the nested object style though, see comment

patriknw commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -108,36 +108,69 @@ akka {
         idle-cpu-level = 5
-        
+
+        system-message-resend-interval = 1 second
```
# unacknowledged system messages are re-delivered with this interval

patriknw commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -108,36 +108,69 @@ akka {
-        
+
+        system-message-resend-interval = 1 second
+        handshake-retry-interval = 1 second
```
`# incomplete handshake attempt is retried with this interval`

patriknw commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -108,36 +108,69 @@ akka {
+
+        system-message-resend-interval = 1 second
+        handshake-retry-interval = 1 second
+        inject-handshake-interval = 1 second
```
```
# handshake requests are performed periodically with this interval, 
# also after the handshake has been completed to be able to establish
# a new session with a restarted destination system
```

patriknw commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -108,36 +108,69 @@ akka {
+        system-message-resend-interval = 1 second
+        handshake-retry-interval = 1 second
+        inject-handshake-interval = 1 second
+        give-up-send-after = 60 seconds
```
`# messages that are not accepted by Aeron are dropped after retrying for this period`

patriknw commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -108,36 +108,69 @@ akka {
+        handshake-retry-interval = 1 second
+        inject-handshake-interval = 1 second
+        give-up-send-after = 60 seconds
+        shutdown-flush-timeout = 1 second
```
`# during ActorSystem termination the remoting will wait this long for a acknowledgment by the destination system that flushing of outstanding remote messages has been completed`

2m committed:
#20587 Fixed compilation errors after rebase and addressed PR feedback

2m committed:
#20587 Copy reused settings from remote to the artery

2m commented:
Thanks for the descriptions of settings without comments. I have rebased and addressed the feedback.

@ktoso I went with no `with...` methods because every setting is either from `Config` or a compile time constant. Therefore we should use the facilities provided by Config library for overrides instead of duplicating the logic in the configuration classes as well.

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/ArterySettings.scala:
```diff
@@ -0,0 +1,103 @@
+    val SysMsgBufferSize: Int = getInt("system-message-buffer-size").requiring(
+      _ > 0, "system-message-buffer-size must be more than zero")
+    val SystemMessageResendInterval = config.getMillisDuration("system-message-resend-interval").requiring(interval ⇒
+      interval > 0.seconds, "system-message-resend-interval must be more than zero")
```
`Duration.Zero`

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/ArterySettings.scala:
```diff
@@ -0,0 +1,103 @@
+      interval > 0.seconds, "handshake-retry-interval must be more than zero")
+    val InjectHandshakeInterval = config.getMillisDuration("inject-handshake-interval").requiring(interval ⇒
+      interval > 0.seconds, "inject-handshake-interval must be more than zero")
+    val GiveUpSendAfter = config.getMillisDuration("give-up-send-after")
```
add requiring > Duration.Zero on all the timeouts

patriknw commented:
LGTM, minor nitpick

patriknw commented:
I created an issue for the test failure: https://github.com/akka/akka/issues/21331
I'm pretty sure it's unrelated.

patriknw commented:
PLS BUILD

patriknw commented:
merging, so I can rebase my PR and add my config without too much conflicts
will fix the nitpicks in my PR
